### PR TITLE
Feature/helm chart

### PR DESCRIPTION
**Pull request by ganievs**
Add the ability to generate the operator's chart from manifests

### CODE REVIEW

**Comment by alexandrevilain** [comment]
Thanks for your contribution! 
That's a really great idea! 
Don't you think it could be better to run the helmify command only one time and push the first version of the helm chart in the PR ?

Then we'll have to:
- update the "prepare-release" target to update helm chart
- find a way to host charts maybe using github pages
- update the release workflow to push the helm chart for each new releases.

What do you think ?

File: charts/temporal-operator/Chart.yaml [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+type: application
+# This is the chart version. This version number should be incremented each time you make changes
+# to the chart and its templates, including the app version.
+# Versions are expected to follow Semantic Versioning (https://semver.org/)
+version: 0.1.0
+# This is the version number of the application being deployed. This version number should be
+# incremented each time you make changes to the application. Versions are not expected to
+# follow Semantic Versioning. They should reflect the version the application is using.
+# It is recommended to use it with quotes.
+appVersion: "0.1.0"
```
Comment (alexandrevilain) [inline charts/temporal-operator/Chart.yaml:21]: I think it's a best practice to set the appVersion to the operator's version. Could you please set it to `0.15.0` ?

File: charts/temporal-operator/Chart.yaml [hunk 2]
```diff
@@ -0,0 +1,21 @@
+apiVersion: v2
+name: temporal-operator
+description: A Helm chart for Kubernetes
```
Comment (alexandrevilain) [inline charts/temporal-operator/Chart.yaml:3]:
```suggestion
description: The Temporal Kubernetes operator helm chart
```

File: charts/temporal-operator/templates/deployment.yaml [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+      {{- include "temporal-operator.selectorLabels" . | nindent 8 }}
+      annotations:
+        kubectl.kubernetes.io/default-container: manager
+    spec:
+      containers:
+      - args: {{- toYaml .Values.controllerManager.manager.args | nindent 8 }}
+        command:
+        - /manager
+        env:
+        - name: KUBERNETES_CLUSTER_DOMAIN
```
Comment (alexandrevilain) [inline charts/temporal-operator/templates/deployment.yaml:27]: This environment variable isn't used by the operator

File: charts/temporal-operator/templates/selfsigned-issuer.yaml [hunk 4]
```diff
@@ -0,0 +1,8 @@
+apiVersion: cert-manager.io/v1
```
Comment (alexandrevilain) [inline charts/temporal-operator/templates/selfsigned-issuer.yaml:1]: I think we could add a condition to create this issuer or not.
Users can choose to not use cert-manager Issuer to create webhook's certs

File: charts/temporal-operator/templates/serving-cert.yaml [hunk 5]
```diff
@@ -0,0 +1,16 @@
+apiVersion: cert-manager.io/v1
```
Comment (alexandrevilain) [inline charts/temporal-operator/templates/serving-cert.yaml:1]: Same comment as for issuer

File: charts/temporal-operator/values.yaml [hunk 6]
```diff
@@ -0,0 +1,27 @@
+controllerManager:
```
Comment (alexandrevilain) [inline charts/temporal-operator/values.yaml:1]: IMO we can remove the `controllerManager` prefix. Looks `manager` is sufficient.

**Review by alexandrevilain — CHANGES_REQUESTED** [review]

File: charts/temporal-operator/templates/deployment.yaml (hunk 3 above)
Comment (ganievs) [reply charts/temporal-operator/templates/deployment.yaml]: Removed

File: charts/temporal-operator/values.yaml (hunk 6 above)
Comment (ganievs) [reply charts/temporal-operator/values.yaml]: Renamed `controllerManager.manager` -> `manager`

File: charts/temporal-operator/templates/selfsigned-issuer.yaml (hunk 4 above)
Comment (ganievs) [reply charts/temporal-operator/templates/selfsigned-issuer.yaml:2]: Add `certissue: true` in the values file to make it optional.

File: charts/temporal-operator/templates/serving-cert.yaml (hunk 5 above)
Comment (ganievs) [reply charts/temporal-operator/templates/serving-cert.yaml:2]: `certissue: true` also makes it optional.

File: charts/temporal-operator/Chart.yaml (hunk 2 above)
Comment (ganievs) [reply charts/temporal-operator/Chart.yaml]: Updated

File: charts/temporal-operator/Chart.yaml (hunk 1 above)
Comment (ganievs) [reply charts/temporal-operator/Chart.yaml]: The version is 0.15.0 now

File: Makefile [hunk 7]
```diff
@@ -169,6 +169,11 @@ artifacts: kustomize
 	$(KUSTOMIZE) build config/crd > ${RELEASE_PATH}/temporal-operator.crds.yaml
 	$(KUSTOMIZE) build config/default > ${RELEASE_PATH}/temporal-operator.yaml
 
+.PHONY: artifacts
```
Comment (alexandrevilain) [inline Makefile:172]:
```suggestion
.PHONY: helm
```

File: Makefile [hunk 8]
```diff
@@ -169,6 +169,11 @@ artifacts: kustomize
...
+helm: manifests helmify
```
Comment (alexandrevilain) [inline Makefile:173]: As you've overwritten the generated chart, are the new Makefile targets still useful ? IMO we can remove them. 
Helmify was just a bootstraping tool, am I wrong ?

File: charts/temporal-operator/Chart.yaml [hunk 9]
```diff
@@ -0,0 +1,21 @@
+apiVersion: v2
+name: temporal-operator
+description: he Temporal Kubernetes operator helm chart
```
Comment (alexandrevilain) [inline charts/temporal-operator/Chart.yaml:3]:
```suggestion
description: The Temporal Kubernetes operator helm chart
```

File: Makefile (hunk 8 above)
Comment (alexandrevilain) [reply Makefile:173]: Hum this target could still be revelent to push new CRDS to the helm chart when preparing release ?

File: charts/temporal-operator/templates/selfsigned-issuer.yaml (hunk 4 above)
Comment (alexandrevilain) [reply charts/temporal-operator/templates/selfsigned-issuer.yaml:2]: Why `and` ? Isn't `{{- if .Values.certissue }}` enough ?

Comment (alexandrevilain) [reply charts/temporal-operator/templates/selfsigned-issuer.yaml:2]: What do you think about `.Values.webhook.certManager.selfSignedIssuer.enabled` ?

File: charts/temporal-operator/templates/serving-cert.yaml (hunk 5 above)
Comment (alexandrevilain) [reply charts/temporal-operator/templates/serving-cert.yaml:2]: What do you think about `.Values.webhook.certManager.certificate.enabled` ? with a key to specify which kind of issuer we want to use ?

File: charts/temporal-operator/values.yaml [hunk 10]
```diff
@@ -0,0 +1,27 @@
...
+    requests:
+      cpu: 10m
+      memory: 64Mi
+  replicas: 1
+  serviceAccount:
+    annotations: {}
+certissue: true
+imagePullSecrets: []
+kubernetesClusterDomain: cluster.local
+webhookService:
```
Comment (alexandrevilain) [inline charts/temporal-operator/values.yaml:22]:
```suggestion
webhook:
  service:
```
What do you think ?

**Review by alexandrevilain — CHANGES_REQUESTED** [review]
Thanks for the update @ganievs !
I still have some comments but it starts to become really cool!

**Comment by alexandrevilain** [comment]
Once everything will be ready, could you please squash your commits ? Thanks :)

File: charts/temporal-operator/values.yaml (hunk 10 above)
Comment (ganievs) [reply charts/temporal-operator/values.yaml]: Agree. Rendamed

File: Makefile (hunk 7 above)
Comment (ganievs) [reply Makefile:172]: renamed

File: Makefile (hunk 8 above)
Comment (ganievs) [reply Makefile:173]: Seems the helmify is unnecessary in this case. Enough to copy `${RELEASE_PATH}/temporal-operator.crds.yaml` to the chart's crds directory.

File: charts/temporal-operator/templates/selfsigned-issuer.yaml (hunk 4 above)
Comment (ganievs) [reply charts/temporal-operator/templates/selfsigned-issuer.yaml]: Good point and what do you think if certissue values will look like this? 
```
  certManager:
    certificate:
      enabled: true
      useCustomIssuer: false
      # Issuer references if you want to use custom issuer
      # In other case will be used selfSigned issuer
      issuerRef: {}
```

**Comment by alexandrevilain** [comment]
Hi @ganievs !
Something went wrong with your squash ? I see only two files to review now

**Comment by ganievs** [comment]
> Hi @ganievs ! Something went wrong with your squash ? I see only two files to review now

Hi, @alexandrevilain sorry for the inconvenience. I forgot to add the chart directory.

**Comment by bjet007** [comment]
Hi @ganievs, thank's for the contribution. I did the the helm chart with 0.15.4 and before merging the CRD should be updated.

**Comment by ganievs** [comment]
HI @bjet007! Done.

**Comment by bjet007** [comment]
Hi @ganievs, since 0.16.0 is out, I guess the MR should be updated again to match the CRD?  

@alexandrevilain do you have an ETA to merge it back into master and be published at the same time than your release?

File: charts/temporal-operator/values.yaml [hunk 11]
```diff
@@ -0,0 +1,35 @@
+manager:
+  args:
+  - --leader-elect
+  containerSecurityContext:
+    allowPrivilegeEscalation: false
+  image:
+    repository: ghcr.io/alexandrevilain/temporal-operator
+    tag: v0.13.3
```
Comment (bjet007) [inline charts/temporal-operator/values.yaml:8]: The tag here does not match the appVersion bumped to v0.15.4

**Comment by alexandrevilain** [comment]
> @alexandrevilain do you have an ETA to merge it back into master and be published at the same time than your release?

Yes sorry, it went under the radar. I have to double check for this PR content again and merge it asap.

**Comment by bjet007** [comment]
> > @alexandrevilain do you have an ETA to merge it back into master and be published at the same time than your release?
> 
> Yes sorry, it went under the radar. I have to double check for this PR content again and merge it asap.

No problem, found a small issue with the image tag not matching the chart appVersion which lead to running 0.13.0 operator while chart is 0.15.4.

**Comment by ganievs** [comment]
> > > @alexandrevilain do you have an ETA to merge it back into master and be published at the same time than your release?
> > 
> > 
> > Yes sorry, it went under the radar. I have to double check for this PR content again and merge it asap.
> 
> No problem, found a small issue with the image tag not matching the chart appVersion which lead to running 0.13.0 operator while chart is 0.15.4.

Removed `tag` definition from values.yaml. Now image tag is the same as appVersion by default

**Comment by alexandrevilain** [comment]
Hi @ganievs !
Could you please rebase and squash your commits ? 
Then I'll be able to review your PR :)

**Comment by ganievs** [comment]
> Hi @ganievs ! Could you please rebase and squash your commits ? Then I'll be able to review your PR :)

done

File: charts/temporal-operator/templates/deployment.yaml [hunk 12]
```diff
@@ -0,0 +1,59 @@
...
+        control-plane: controller-manager
+      {{- include "temporal-operator.selectorLabels" . | nindent 8 }}
+      annotations:
+        kubectl.kubernetes.io/default-container: manager
+    spec:
+      containers:
+      - args: {{- toYaml .Values.manager.args | nindent 8 }}
+        command:
+        - /manager
+        image: {{ .Values.manager.image.repository }}:{{ .Values.manager.image.tag | default .Chart.AppVersion }}
```
[the remark this answers is not in the document]
Comment (ganievs) [reply charts/temporal-operator/templates/deployment.yaml:26]: Hi @Nepoxx! Thanks for highlight, fixed

**Review by alexandrevilain — APPROVED** [review]

**Comment by alexandrevilain** [comment]
Thanks @ganievs ! This PR is now approved ! Waiting for the CI to pass. 
If I remember well this PR provided github actions too ? I can't see it right now, is this a rebase issue ?

**Comment by ElanHasson** [comment]
Congratulations on this great milestone!!

**Pull request merged**